### PR TITLE
MAC API change: additional arguments to EVP_MAC_init()

### PR DESCRIPTION
**Pull request by paulidale**
Adding the key, key length and params arguments to the MAC init() call.

Refer to #11007

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by paulidale** [comment]
Also fixing a regression in the SipHash API where the numbers of rounds could no longer be specified.

File: doc/man3/EVP_MAC.pod [hunk 1]
```diff
@@ -114,7 +115,8 @@ I<ctx>.
 =head2 Computing functions
 
 EVP_MAC_init() sets up the underlying context with information given
-through diverse controls.
+via the I<key> and I<params> arguments.  The MAC I<key> has a length of
+I<keylen> and the parameters in I<params> are processed before setting the key.
```
Comment (slontis) [inline doc/man3/EVP_MAC.pod:119]: Say something about what happens if the  key is NULL.

File: doc/man7/EVP_MAC-KMAC.pod [hunk 2]
```diff
@@ -47,7 +47,7 @@ The default value is 0.
 
 =back
 
-The "custom" and "key" parameters must be set before EVP_MAC_init().
+The "custom" parameter must be set before EVP_MAC_init().
```
Comment (slontis) [inline doc/man7/EVP_MAC-KMAC.pod:50]: The key parameter must be set before or during the init()?

File: doc/man7/provider-mac.pod [hunk 3]
```diff
@@ -108,7 +109,8 @@ I<mctx> parameter and return the duplicate copy.
 =head2 Encryption/Decryption Functions
 
 OSSL_FUNC_mac_init() initialises a mac operation given a newly created provider
-side mac context in the I<mctx> parameter.
+side mac context in the I<mctx> parameter.  The I<params> are set before setting
+the MAC I<key> of I<keylen> bytes.
```
Comment (slontis) [inline doc/man7/provider-mac.pod:113]: I wonder if there is any case that relies on the keylength being known before setting a param?

File: providers/implementations/kdfs/sskdf.c [hunk 4]
```diff
@@ -223,27 +223,19 @@ static int SSKDF_mac_kdm(EVP_MAC_CTX *ctx_init,
...
-        goto end;
-
     if (!kmac_init(ctx_init, kmac_custom, kmac_custom_len, kmac_out_len,
                    derived_key_len, &kmac_buffer))
         goto end;
     if (kmac_buffer != NULL)
         mac = kmac_buffer;
 
-    if (!EVP_MAC_init(ctx_init))
+    if (!EVP_MAC_init(ctx_init, salt, salt_len, NULL))
```
Comment (slontis) [inline providers/implementations/kdfs/sskdf.c:238]: hopefully this ordering change doesnt break anything

File: providers/implementations/macs/blake2_mac_impl.c [hunk 5]
```diff
@@ -87,15 +87,25 @@ static size_t blake2_mac_size(void *vmacctx)
...
-    /* Check key has been set */
-    if (macctx->params.key_length == 0) {
+    if (key != NULL) {
+        if (keylen > BLAKE2_KEYBYTES || keylen == 0) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_KEY_LENGTH);
+            return 0;
+        }
+        memcpy(macctx->key, key, keylen);
+        /* Pad with zeroes at the end if required */
+        if (keylen < BLAKE2_KEYBYTES)
```
Comment (slontis) [inline providers/implementations/macs/blake2_mac_impl.c:104]: Not sure if there is a way of sharing this block with the settable case..

File: providers/implementations/macs/hmac_prov.c [hunk 6]
```diff
@@ -141,19 +141,31 @@ static size_t hmac_size(void *vmacctx)
...
 {
     struct hmac_data_st *macctx = vmacctx;
     const EVP_MD *digest;
     int rv = 1;
 
-    if (!ossl_prov_is_running())
+    if (!ossl_prov_is_running() || !hmac_set_ctx_params(macctx, params))
         return 0;
 
+    if (key != NULL) {
```
Comment (slontis) [inline providers/implementations/macs/hmac_prov.c:154]: I dont think I like this duping of code..

**Review by slontis — COMMENTED** [review]
I would prefer to see some setkey() like function if possible that can be shared by the init() and settable cases..

File: providers/implementations/macs/siphash_prov.c [hunk 7]
```diff
@@ -117,52 +139,71 @@ static int siphash_final(void *vmacctx, unsigned char *out, size_t *outl,
...
             return 0;
     }
     if ((p = OSSL_PARAM_locate_const(params, OSSL_MAC_PARAM_KEY)) != NULL)
         if (p->data_type != OSSL_PARAM_OCTET_STRING
             || p->data_size != SIPHASH_KEY_SIZE
-            || !SipHash_Init(&ctx->siphash, p->data, 0, 0))
+            || !SipHash_Init(&ctx->siphash, p->data,
+                             crounds(ctx), drounds(ctx)))
             return 0;
+    if ((p = OSSL_PARAM_locate_const(params, OSSL_MAC_PARAM_C_ROUNDS)) != NULL
```
Comment (slontis) [inline providers/implementations/macs/siphash_prov.c:201]: these should moved up before the key locate so that they get set and passed to the SipHash_Init()

File: test/recipes/30-test_evp_data/evpmac_siphash.txt [hunk 8]
```diff
@@ -155,7 +155,7 @@ Output = 5150d1772f50834a503e069a973fbd7c
 MAC = SipHash
 Ctrl = size:13
 Key = 000102030405060708090A0B0C0D0E0F
-Result = MAC_BAD_PARAMS
+Result = MAC_INIT_ERROR
 
 # SIPHASH - default values: 2,4 rounds, explicit 13-byte mac (invalid size)
```
Comment (slontis) [inline test/recipes/30-test_evp_data/evpmac_siphash.txt:160]: Add a test that sets the rounds (for the case you just added)

File: doc/man3/EVP_MAC.pod (hunk 1 above)
Comment (paulidale) [reply doc/man3/EVP_MAC.pod:119]: Okay, will add.

File: doc/man7/provider-mac.pod (hunk 3 above)
Comment (paulidale) [reply doc/man7/provider-mac.pod:113]: Not yet :)

It's still possible to set params after the key.  It just can't be done using the `init()` call.  I don't think this is a problem.

File: providers/implementations/macs/siphash_prov.c (hunk 7 above)
Comment (paulidale) [reply providers/implementations/macs/siphash_prov.c:201]: Good catch.

**Comment by paulidale** [comment]
> I would prefer to see some setkey() like function if possible that can be shared by the init() and settable cases..

I was wondering if the set_params path to set the key was needed or not.  It would require some additional rework to omit in a few places.

File: doc/man7/provider-mac.pod [hunk 9]
```diff
@@ -153,7 +155,8 @@ parameters are relevant to, or are understood by all macs:
 
 =item "key" (B<OSSL_MAC_PARAM_KEY>) <octet string>
 
-Sets the key in the associated MAC ctx.
+Sets the key in the associated MAC ctx.  This is identical to passing a I<key>
+argument to the EVP_MAC_init() function.
```
Comment (levitte) [inline doc/man7/provider-mac.pod:159]: If we're talking in terms of this manual page, shouldn't that refer to `OSSL_FUNC_mac_init()`?

File: providers/implementations/macs/hmac_prov.c (hunk 6 above)
Comment (levitte) [reply providers/implementations/macs/hmac_prov.c:154]: Me neither.  @paulidale, please make a separate function that treats key and key length appropriately, and that gets called both from here and from the set_params function.

This goes for the same sort of code copying in the other implementations

File: doc/man7/provider-mac.pod (hunk 3 above)
Comment (levitte) [reply doc/man7/provider-mac.pod:113]: How do you know in what order the arguments are treated?  You know what *our* implementations do, but you have no idea what others will do.  Some might choose to treat *some* of the params before the key and keylength, and *some* after.  It really shouldn't matter, as long as they do something *sensible*, and that's up to each provider author to figure out.

File: providers/implementations/macs/blake2_mac_impl.c (hunk 5 above)
Comment (paulidale) [reply providers/implementations/macs/blake2_mac_impl.c:104]: There is.

File: doc/man7/provider-mac.pod (hunk 9 above)
Comment (paulidale) [reply doc/man7/provider-mac.pod:159]: Yep.  Changed.

File: providers/implementations/macs/hmac_prov.c (hunk 6 above)
Comment (levitte) [reply providers/implementations/macs/hmac_prov.c]: I see that this was already mentioned elsewhere

File: test/recipes/30-test_evp_data/evpmac_siphash.txt (hunk 8 above)
Comment (paulidale) [reply test/recipes/30-test_evp_data/evpmac_siphash.txt:160]: Added.

File: providers/implementations/macs/hmac_prov.c (hunk 6 above)
Comment (paulidale) [reply providers/implementations/macs/hmac_prov.c]: Done it for all the MACs.

**Comment by paulidale** [comment]
Feedback addressed.

**Comment by levitte** [comment]
> I was wondering if the set_params path to set the key was needed or not. It would require some additional rework to omit in a few places.

I think the answer to that should be pretty obvious, and is the actual reason why it's an OSSL_PARAM to begin with.  From `man openssl dgst`:

> **-macopt nm:v**
> Passes options to MAC algorithm, specified by -mac key.  Following
> options are supported by both by HMAC and gost-mac:
> 
> **key:string**
> Specifies MAC key as alphanumeric string (use if key contain
> printable characters only). String length must conform to any
> restrictions of the MAC algorithm for example exactly 32 chars
> for gost-mac.
> 
> **hexkey:string**
> Specifies MAC key in hexadecimal form (two hex digits per
> byte).  Key length must conform to any restrictions of the MAC
> algorithm for example exactly 32 chars for gost-mac.

**Comment by paulidale** [comment]
Also added an OSSL_PARAM argument to the KDF derive call.

**Comment by paulidale** [comment]
Add an OSSL_PARAM array argument to the instantiate call for RNGs.

Reviews please.

**Comment by slontis** [comment]
needs a rebase.

**Review by slontis — COMMENTED** [review]
Is  setparams revelant anymore for kdf if both init and derive pass params?

File: crypto/crmf/crmf_pbm.c [hunk 10]
```diff
@@ -207,12 +207,10 @@ int OSSL_CRMF_pbm_new(OSSL_LIB_CTX *libctx, const char *propq,
 
     macparams[0] = OSSL_PARAM_construct_utf8_string(OSSL_MAC_PARAM_DIGEST,
                                                     (char *)mdname, 0);
-    macparams[1] = OSSL_PARAM_construct_octet_string(OSSL_MAC_PARAM_KEY,
-                                                     basekey, bklen);
     if ((mac = EVP_MAC_fetch(libctx, "HMAC", propq)) == NULL
             || (mctx = EVP_MAC_CTX_new(mac)) == NULL
             || !EVP_MAC_CTX_set_params(mctx, macparams)
```
Comment (slontis) [inline crypto/crmf/crmf_pbm.c:212]: this set_params is not needed..

**Comment by paulidale** [comment]
Rebased.

Setparams is still used for KDFs from the TLS stack.
It might be possible to figure out a way to avoid this but is it useful?

The larger question is _is setparams useful for any of these algorithms anymore?_  I think it's worth including, if for no other reason than consistency across the APIs.

File: doc/man3/EVP_MAC.pod [hunk 11]
```diff
@@ -354,6 +357,7 @@ EVP_MAC_do_all_provided() returns nothing at all.
       const char *cipher = getenv("MY_MAC_CIPHER");
       const char *digest = getenv("MY_MAC_DIGEST");
       const char *key = getenv("MY_KEY");
+      size_t keylen = strlen(key);
```
Comment (slontis) [inline doc/man3/EVP_MAC.pod:360]: not a good pattern to do strlen(key) and then check key == NULL further down..

File: doc/man7/EVP_MAC-KMAC.pod [hunk 12]
```diff
@@ -47,8 +47,10 @@ The default value is 0.
...
```
Comment (slontis) [inline doc/man7/EVP_MAC-KMAC.pod:50]: This line isnt quite right either. Since custom could now be set in the init..

File: doc/man3/EVP_MAC.pod (hunk 11 above)
Comment (paulidale) [reply doc/man3/EVP_MAC.pod:360]: Good point.
Fixed.

File: providers/implementations/kdfs/kbkdf.c [hunk 13]
```diff
@@ -330,16 +330,9 @@ static int kbkdf_set_ctx_params(void *vctx, const OSSL_PARAM params[])
...
     /* Set up digest context, if we can. */
-    if (ctx->ctx_init != NULL && ctx->ki_len != 0) {
-        mparams[0] = OSSL_PARAM_construct_octet_string(OSSL_MAC_PARAM_KEY,
-                                                       ctx->ki, ctx->ki_len);
-        mparams[1] = OSSL_PARAM_construct_end();
-
-        if (!EVP_MAC_CTX_set_params(ctx->ctx_init, mparams)
-            || !EVP_MAC_init(ctx->ctx_init))
+    if (ctx->ctx_init != NULL && ctx->ki_len != 0
+            && !EVP_MAC_init(ctx->ctx_init, ctx->ki, ctx->ki_len, NULL))
```
Comment (slontis) [inline providers/implementations/kdfs/kbkdf.c:334]: Why has this disappeared?

File: doc/man7/EVP_MAC-KMAC.pod (hunk 12 above)
Comment (paulidale) [reply doc/man7/EVP_MAC-KMAC.pod:50]: Good point again.

File: providers/implementations/kdfs/kbkdf.c (hunk 13 above)
Comment (paulidale) [reply providers/implementations/kdfs/kbkdf.c:334]: It was `mparams` which contained only the key.  Since the key is a function argument, there is no point passing the empty params.

File: providers/implementations/rands/drbg.c [hunk 14]
```diff
@@ -365,9 +365,6 @@ int ossl_prov_drbg_instantiate(PROV_DRBG *drbg, unsigned int strength,
     size_t noncelen = 0, entropylen = 0;
     size_t min_entropy, min_entropylen, max_entropylen;
 
-    if (!ossl_prov_is_running())
```
Comment (slontis) [inline providers/implementations/rands/drbg.c:368]: Why was this removed??

Comment (slontis) [reply providers/implementations/rands/drbg.c:368]: The one in rand_drbg_restart() is ok.. So as long as the instatiate wrapper ones do the check then this is ok,.

Comment (paulidale) [reply providers/implementations/rands/drbg.c:368]: I added this check to the three DRBGs along with the set_params call.

I contemplated leaving it there to avoid this question it but figured it was better to do the job properly 😃

Comment (slontis) [reply providers/implementations/rands/drbg.c:368]: Maybe just keep in mind if this code 'changes'.

**Review by slontis — APPROVED** [review]
reword one of your commits that has 'DERBG' instantiate, unless you can explain what the E is :)

**Comment by paulidale** [comment]
> reword one of your commits that has 'DERBG' instantiate, unless you can explain what the E is :)

Clearly it's DEterministic :)

**Comment by slontis** [comment]
> Clearly it's DEterministic :)

I think it might be that E is adjacent to both D and R on the keyboard....

**Comment by paulidale** [comment]
Stop ruining the plot :)

File: doc/man7/provider-mac.pod [hunk 15]
```diff
@@ -211,6 +214,16 @@ With some MAC algorithms, it can also be used to set the size that the
...
+=item "c-rounds" (B<OSSL_MAC_PARAM_C_ROUNDS>) <unsigned integer>
+
+Specifies the number of rounds per message block for SipHash.  By default this
+is I<2>.
+
+=item "d-rounds" (B<OSSL_MAC_PARAM_D_ROUNDS>) <unsigned integer>
+
+Specifies the number of finalisation rounds for SipHash.  By default this is
+I<4>.
+
```
Comment (levitte) [inline doc/man7/provider-mac.pod:226]: Why are you adding these here, especially since you already added them in EVP_MAC-SipHash.pod???

**Review by levitte — CHANGES_REQUESTED** [review]

File: doc/man7/provider-mac.pod (hunk 15 above)
Comment (paulidale) [reply doc/man7/provider-mac.pod:226]: Should we not document this?

Comment (levitte) [reply doc/man7/provider-mac.pod:226]: Yes.  Which you have done in EVP_MAC-SipHash.pod.  This is a misplaced duplicate

Comment (paulidale) [reply doc/man7/provider-mac.pod:226]: There are plenty more cases of similar.  I'm happy to remove this but it goes against the grain.

Comment (paulidale) [reply doc/man7/provider-mac.pod:226]: We document _key_, _IV_, _custom_, _salt_, _block XOF_, _digest no-init_, _digest-one-shot_ in this man page.  Why not _crounds_ and _drounds_?

Please elucidate as to a consequential difference between the former and the latter?

Comment (levitte) [reply doc/man7/provider-mac.pod:226]: I actually think those are generally misplaced in this particular file, *unless* they are parameters that are expected to be handled by *all* MAC implementations without exception.

The idea with `provider-mac.pod` is to be a manual for *provider authors* (there's even a note about that at the beginning of DESCRIPTION), and are therefore meant to be generic rather than a description of *our* providers, even to the point of not describing individual implementations.

The idea is that *our* providers and *our* implementations are described in other manuals dedicated to them.  So the description of *our* default provider is in `OSSL_PROVIDER-default.pod`, and the description of *our* SipHash implementation is in `EVP_MAC-SipHash.pod`.

In essence, it's about very different audiences:
-   `provider-{name}.pod` are for provider *authors*
-   `{METHODNAME}-[AlgorithmName}.pod` are for provider *users*

[I have said this a number of times]

Comment (paulidale) [reply doc/man7/provider-mac.pod]: I've removed these two from the file.

**Comment by paulidale** [comment]
Merged to master.  Thanks for the feedback.
